# Question
# 1. -> Gender Classification using the given dataset

In [1]:
# Gender is my target variable in this dataset

In [33]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [52]:
# from google.colab import files
# uploaded = files.upload()

In [35]:
data = pd.read_csv('/content/Information.csv', encoding='latin1')

In [36]:
data.head()

_unit_id  _golden  ...   tweet_location               user_timezone
0  815719226    False  ...  main; @Kan1shk3                     Chennai
1  815719227    False  ...              NaN  Eastern Time (US & Canada)
2  815719228    False  ...           clcncl                    Belgrade
3  815719229    False  ...    Palo Alto, CA  Pacific Time (US & Canada)
4  815719230    False  ...              NaN                         NaN

[5 rows x 26 columns]

In [37]:
import numpy as np 
from sklearn.naive_bayes import MultinomialNB

import nltk 
import re

In [38]:
data = pd.concat([data.gender, data.description], axis=1)
data.dropna(axis = 0, inplace = True)

In [39]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16224 entries, 0 to 20049
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   gender       16224 non-null  object
 1   description  16224 non-null  object
dtypes: object(2)
memory usage: 380.2+ KB


In [40]:
import nltk # natural language took kit
nltk.download("stopwords")
nltk.download('punkt')
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [41]:
import nltk
description_list = [] # we created a list so we after these steps, we will append into this list
for description in data.description:
    description = re.sub("[^a-zA-Z]", " ", description)
    # sub method finds the given pattern ([^a-zA-Z] means, NOT letter like ":") and changes them with " " (space)
    description = description.lower()
    # we need to have all letters lowercase (because A is not equall to a)
    description = nltk.word_tokenize(description)
    # we make a word list from our text
    lemma = nltk.WordNetLemmatizer()
    description = [lemma.lemmatize(word) for word in description]
    # we found the roots of each words with lemma
    description = " ".join(description)
    # after all these steps,we joined the words together and remake our text.
    description_list.append(description)
    # and append these texts into the list we created.

In [42]:
from sklearn.feature_extraction.text import CountVectorizer

max_features = 5000 

cv = CountVectorizer(max_features=max_features, stop_words = "english")
# in this method, we remove the stopwords (irrelevant words) in English language. (like "of", "and", "the" etc.)

sparce_matrix = cv.fit_transform(description_list).toarray()

print("top used {} words: {}".format(max_features, cv.get_feature_names()))

top used 5000 words: ['aa', 'aaron', 'abc', 'aberdeen', 'ability', 'able', 'absolute', 'absolutely', 'absurdity', 'abuse', 'ac', 'academia', 'academic', 'academy', 'acc', 'accept', 'accepted', 'accepting', 'access', 'accessible', 'accessory', 'accident', 'accommodation', 'accomplish', 'account', 'accountant', 'accounting', 'ace', 'achieve', 'act', 'acting', 'action', 'active', 'activist', 'activity', 'actor', 'actress', 'actual', 'actually', 'ad', 'adalah', 'adam', 'add', 'added', 'addict', 'addicted', 'addiction', 'addition', 'address', 'admin', 'administrator', 'admirer', 'adopted', 'adoption', 'adorable', 'adore', 'adult', 'advance', 'advanced', 'adventure', 'adventurer', 'advertising', 'advice', 'advise', 'adviser', 'advisor', 'advocacy', 'advocate', 'advocating', 'aerial', 'aerospace', 'aesthetic', 'af', 'afc', 'affair', 'affiliate', 'affiliated', 'affiliation', 'affordable', 'aficionado', 'afraid', 'africa', 'african', 'afrikaner', 'afro', 'afternoon', 'ag', 'age', 'agency', 'age

In [43]:
y = data.iloc[:, 0].values
x = sparce_matrix


In [44]:
from sklearn.model_selection import cross_val_score
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn import tree
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC

In [45]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.1, random_state = 42)

In [46]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
rf = RandomForestClassifier()
rf.fit(x_train, y_train)
# prediction
y_pred = rf.predict(x_test)

In [47]:
# Random Forest 
accuracy = 100.0 * accuracy_score(y_test, y_pred)
print("Accuracy: ", accuracy)

Accuracy:  54.46703635243376


In [48]:
from sklearn.naive_bayes import GaussianNB
nb = GaussianNB()
nb.fit(x_train,y_train)
print("accuracy: ",nb.score(x_test,y_test))

accuracy:  0.3659889094269871


In [49]:
from sklearn.metrics import accuracy_score
nb = MultinomialNB()
nb.fit(x_train, y_train)
# prediction
y_pred = nb.predict(x_test)
accuracy = 100.0 * accuracy_score(y_test, y_pred)
print("Accuracy: ", accuracy)

Accuracy:  58.5335797905114


In [50]:
lr = LogisticRegression(max_iter = 2000)
lr.fit(x_train,y_train)
# prediction
y_pred = lr.predict(x_test)
accuracy = 100.0 * accuracy_score(y_test, y_pred)
print("Accuracy: ", accuracy)

Accuracy:  56.13062230437461


# Hence the highest accuracy is 58.5335797905114 by using Multinomial Naive Bias Algorithm

# Question
# 2. -> What are the most common emotions/words used by Males and Females

In [51]:
data = pd.read_csv('/content/Information.csv', encoding='latin1')

In [ ]:
data = pd.concat([data.gender, data.text], axis=1)
data.dropna(axis = 0, inplace = True)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19953 entries, 0 to 20049
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   gender  19953 non-null  object
 1   text    19953 non-null  object
dtypes: object(2)
memory usage: 467.6+ KB


In [28]:
import nltk
text_list = [] # we created a list so we after these steps, we will append into this list
for text in data.text:
    text = re.sub("[^a-zA-Z]", " ", text)
    # sub method finds the given pattern ([^a-zA-Z] means, NOT letter like ":") and changes them with " " (space)
    text = text.lower()
    # for converting upper case letters to lower case letters
    text = nltk.word_tokenize(text)
    # for making a word list from our text
    lemma = nltk.WordNetLemmatizer()
    text = [lemma.lemmatize(word) for word in text]
    # for founding the roots of each words with lemma
    text = " ".join(text)
    # after all these steps,we joined the words together and remake our text.
    text_list.append(text)
    # and append these texts into the list we created.

In [31]:
from sklearn.feature_extraction.text import CountVectorizer

max_features = 10 

cv = CountVectorizer(max_features=max_features, stop_words = "english")
# for removing stopwords

sparce_matrix = cv.fit_transform(text_list).toarray()

print("top used {} words: {}".format(max_features, cv.get_feature_names()))

top used 10 words: ['channel', 'day', 'http', 'just', 'like', 'love', 'make', 'update', 'wa', 'weather']


# ['channel', 'day', 'http', 'just', 'like', 'love', 'make', 'update', 'wa', 'weather']
# The above mention words are the most top 10 emotions/words used by both male and females

# Submitted by Nripesh Sharma
# nripesh.sharma@btech.christuniversity.in